src/components/TextFieldMask/CurrencyMask/index.tsx

```diff
@@ -30,22 +37,54 @@ export const CurrencyMask = ({ label = "" }: Props) => {
           dot = dot + 1
         }
       }
-      return splittedThousand.join("") + "," + nonInteger
+      return "R$ " + splittedThousand.join("") + "," + nonInteger
     } else {
-      return DECIMAL_NUMBER.replace(".", ",")
+      return "R$ " + DECIMAL_NUMBER.replace(".", ",")
+    }
+  }
+
+  const onChange = (event: ChangeEvent<HTMLInputElement>) => {
+    if (event.target.value.length > 23) return event.preventDefault()
+
+    const formattedValue = currencyFormatter(event.target.value)
+    const value = currencyFormatter(event.target.value)
+      .replace(/\./g, "")
+      .replace(",", ".")
+    const floatValue = Number(value)
+
+    !event.target.value ? setValue("") : setValue(formattedValue)
+
+    const mappedEvent = {
+      ...event,
+      target: {
+        ...event.target,
+        value: formattedValue
+      }
     }
+
+    onChangeProp && onChangeProp(mappedEvent)
+    onValueChange &&
+      onValueChange(
+        {
+          value: value,
+          formattedValue: formattedValue,
+          floatValue: floatValue
+        },
+        {
+          source: "event" as SourceType.event
+        }
+      )
   }
 
+  useEffect(() => {
+    setValue(currencyFormatter(String(valueProp)))
+  }, [valueProp])
+
   return (
     <TextField
       value={value}
       onChange={(e) => {
```

Comment:
Se a assinatura do método for idêntica ao do prop você pode passar somente a referência do método para o prop. O nome disso é `point-free style`.

https://itnext.io/whats-point-free-style-in-typescript-39337000c8cb

https://medium.com/dailyjs/functional-js-7-point-free-style-b21a1416ac6a
```suggestion
      onChange={onChange}
```